lib/harmony/harmonytask/taskTypeHandler.go

```diff
@@ -0,0 +1,285 @@
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/lotus/lib/harmony/harmonydb"
+)
+
+var log = logging.Logger("harmonytask")
+
+type taskTypeHandler struct {
+	TaskInterface
+	TaskTypeDetails
+	TaskEngine *TaskEngine
+	Count      atomic.Int32
+}
+
+func (h *taskTypeHandler) AddTask(extra func(TaskID, *harmonydb.Tx) bool) {
+	var tID TaskID
+	did, err := h.TaskEngine.db.BeginTransaction(h.TaskEngine.ctx, func(tx *harmonydb.Tx) bool {
+		// create taskID (from DB)
+		_, err := tx.Exec(`INSERT INTO harmony_task (name, added_by, posted_time) 
+			VALUES ($1, $2, CURRENT_TIMESTAMP) `, h.Name, h.TaskEngine.ownerID)
+		if err != nil {
+			log.Error("Could not insert into harmonyTask", err)
+			return false
+		}
+		err = tx.QueryRow("SELECT id FROM harmony_task ORDER BY update_time DESC LIMIT 1").Scan(&tID)
+		if err != nil {
+			log.Error("Could not select ID: ", err)
+		}
+		return extra(tID, tx)
+	})
+	if err != nil {
+		log.Error(err)
+	}
+	if !did {
+		return
+	}
+
+	if !h.considerWork("adder", []TaskID{tID}) {
+		h.TaskEngine.bump(h.Name) // We can't do it. How about someone else.
+	}
+}
+
+func (h *taskTypeHandler) considerWork(from string, ids []TaskID) (workAccepted bool) {
+top:
+	if len(ids) == 0 {
+		return true // stop looking for takers
+	}
+
+	// 1. Can we do any more of this task type?
+	if h.Max > -1 && int(h.Count.Load()) == h.Max {
+		log.Infow("did not accept task", "name", h.Name, "reason", "at max already")
+		return false
+	}
+
+	h.TaskEngine.workAdderMutex.Lock()
+	defer h.TaskEngine.workAdderMutex.Unlock()
+
+	// 2. Can we do any more work?
+	err := h.AssertMachineHasCapacity()
+	if err != nil {
+		log.Info(err)
+		return false
+	}
+
+	// 3. What does the impl say?
+	tID, err := h.CanAccept(ids)
+	if err != nil {
+		log.Error(err)
+		return false
+	}
+	if tID == nil {
+		log.Infow("did not accept task", "task_id", ids[0], "reason", "CanAccept() refused")
+		return false
+	}
+
+	// 4. Can we claim the work for our hostname?
+	ct, err := h.TaskEngine.db.Exec(h.TaskEngine.ctx, "UPDATE harmony_task SET owner_id=$1 WHERE id=$2 AND owner_id IS NULL", h.TaskEngine.ownerID, *tID)
+	if err != nil {
+		log.Error(err)
+		return false
+	}
+	if ct == 0 {
+		log.Infow("did not accept task", "task_id", strconv.Itoa(int(*tID)), "reason", "already Taken")
+		var tryAgain = make([]TaskID, 0, len(ids)-1)
+		for _, id := range ids {
+			if id != *tID {
+				tryAgain = append(tryAgain, id)
+			}
+		}
+		ids = tryAgain
+		goto top
+	}
+
+	go func() {
+		h.Count.Add(1)
+		log.Infow("Beginning work on Task", "id", *tID, "from", from, "type", h.Name)
+
+		var done bool
+		var doErr error
+		workStart := time.Now()
+
+		defer func() {
+			if r := recover(); r != nil {
+				log.Error("Recovered from a serious error "+
+					"while processing "+h.Name+" task "+strconv.Itoa(int(*tID))+": ", r)
+			}
+			h.Count.Add(-1)
+
+			h.recordCompletion(*tID, workStart, done, doErr)
+			if done {
+				h.triggerCompletionListeners(*tID)
```

Comment:
So if I'm reading this correctly, the case where the process dies between `recordCompletion` and `triggerCompletionListeners` is handled by other processes looking at harmony_task_history and figuring out what follows to trigger right?

Reply:
Correct. There's always the poller in the ugliest of cases.